### PR TITLE
CATROID-523 Refactor textwatchers of dialog

### PR DESCRIPTION
**Pull request by Walcho1125**
*Please enter a short description of your pull request and add a reference to the Jira ticket.*
Refactor TextWatchers of Dialog, so there is no duplicated code and it also works for the input in "Your Bricks"
https://jira.catrob.at/browse/CATROID-523
### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/AddUserDataToUserDefinedBrickFragment.java [hunk 1]
```diff
@@ -95,8 +95,32 @@ public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
 		TextView addUserDataUserBrickTextView = view.findViewById(R.id.brick_user_defined_add_user_data_description);
 
 		addUserDataUserBrickEditText.setText(userBrickTextView.getText());
+		InputWatcher.TextWatcher textWatcher = new InputWatcher.TextWatcher() {
+			@Override
+			protected boolean isNameUnique(String name) {
```
Comment (mstoeg) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/AddUserDataToUserDefinedBrickFragment.java:100]: I think you don't need to override this function, as the same functionality is already provided by the base class.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/textwatcher/DuplicateInputTextwatcher.java [hunk 2]
```diff
@@ -23,30 +23,27 @@
...
 
-import android.content.Context;
+import org.catrobat.catroid.common.Nameable;
 
-import org.catrobat.catroid.R;
-import org.catrobat.catroid.ui.recyclerview.dialog.TextInputDialog;
+import java.util.List;
 
-import androidx.annotation.Nullable;
+public class DuplicateInputTextwatcher<T extends Nameable> extends InputWatcher.TextWatcher {
```
Comment (mstoeg) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/textwatcher/DuplicateInputTextwatcher.java:30]: Could you please use camelCase for TextWatcher to stay consistent here.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/textwatcher/DuplicateInputTextwatcher.java [hunk 3]
```diff
@@ -23,30 +23,27 @@
...
-			error = context.getString(R.string.name_consists_of_spaces_only);
+	public DuplicateInputTextwatcher(List<T> originalScope) {
+		this.originalScope = originalScope;
+		getScopeFromoriginalScope();
+	}
+	public void setOriginalScope(List<T> scope) {
+		this.originalScope = scope;
+		getScopeFromoriginalScope();
+	}
+	private void getScopeFromoriginalScope() {
```
Comment (mstoeg) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/dialog/textwatcher/DuplicateInputTextwatcher.java:42]: Also please use camelCase here for OriginalScope.

File: catroid/src/main/java/org/catrobat/catroid/merge/NewProjectNameTextWatcher.java [hunk 4]
```diff
@@ -22,34 +22,14 @@
...
-
-		if (input.isEmpty()) {
-			error = context.getString(R.string.name_consists_of_spaces_only);
-		} else if (ReplaceExistingProjectDialogFragment.projectExistsInDirectory(input)) {
-			error = context.getString(R.string.name_already_exists);
-		}
-
-		return error;
+	protected boolean isNameUnique(String name) {
+		return ReplaceExistingProjectDialogFragment.projectExistsInDirectory(name);
```
Comment (mstoeg) [inline catroid/src/main/java/org/catrobat/catroid/merge/NewProjectNameTextWatcher.java:33]: I think there is some kind of logic inversion here, which also leads to the failing testcases. If a project does exist it is NOT unique.

**Review by mstoeg — CHANGES_REQUESTED** [review]
Please have a look at my remarks and also at the two testcases failing in MergeProjectNameTest.

**Comment by Walcho1125** [comment]
@mstoeg Thanks for reviewing, all requested changes should be applied now

**Review by mstoeg — APPROVED** [review]
Well done - looks good now!

**Comment by wslany** [comment]
Thanks!

**Pull request merged**